substrate/frame/delegated-staking/src/lib.rs

```diff
@@ -0,0 +1,820 @@
+		<Delegatees<T>>::get(who)
+			.map_or_else(|| 0u32.into(), |register| register.delegated_balance())
+	}
+
+	fn unbonded_balance(who: &Self::AccountId) -> Self::Balance {
+		<Delegatees<T>>::get(who)
+			.map_or_else(|| 0u32.into(), |register| register.unbonded_balance())
+	}
+
+	fn accept_delegations(
+		who: &Self::AccountId,
+		reward_destination: &Self::AccountId,
+	) -> DispatchResult {
+		// Existing delegatee cannot register again.
+		ensure!(!<Delegatees<T>>::contains_key(who), Error::<T>::NotAllowed);
+
+		// A delegator cannot become a delegatee.
+		ensure!(!<Delegators<T>>::contains_key(who), Error::<T>::NotAllowed);
+
+		// payee account cannot be same as delegatee
+		ensure!(reward_destination != who, Error::<T>::InvalidRewardDestination);
+
+		// make sure they are not already a direct staker or they are migrating.
+		ensure!(
+			T::CoreStaking::status(who).is_err() || <DelegateeMigration<T>>::contains_key(who),
+			Error::<T>::AlreadyStaker
+		);
+
+		// already checked delegatees exist
+		<Delegatees<T>>::insert(
+			who,
+			DelegationLedger {
+				payee: reward_destination.clone(),
+				total_delegated: Zero::zero(),
+				hold: Zero::zero(),
+				pending_slash: Zero::zero(),
+				blocked: false,
+			},
+		);
+
+		Ok(())
+	}
+
+	/// Transfers funds from current staked account to `proxy_delegator`. Current staked account
+	/// becomes a delegatee with `proxy_delegator` delegating stakes to it.
+	fn migrate_accept_delegations(
+		new_delegatee: &Self::AccountId,
+		proxy_delegator: &Self::AccountId,
+		payee: &Self::AccountId,
+	) -> DispatchResult {
+		ensure!(new_delegatee != proxy_delegator, Error::<T>::InvalidDelegation);
+
+		// ensure proxy delegator has at least minimum balance to keep the account alive.
+		ensure!(
+			T::Currency::reducible_balance(
+				proxy_delegator,
+				Preservation::Expendable,
+				Fortitude::Polite
+			) > Zero::zero(),
+			Error::<T>::NotEnoughFunds
+		);
+
+		// ensure staker is a nominator
+		let status = T::CoreStaking::status(new_delegatee)?;
+		match status {
+			StakerStatus::Nominator(_) => (),
+			_ => return Err(Error::<T>::InvalidDelegation.into()),
+		}
+
+		<DelegateeMigration<T>>::insert(&new_delegatee, &proxy_delegator);
+		let stake = T::CoreStaking::stake(new_delegatee)?;
+
+		// unlock funds from staker
+		T::CoreStaking::release_all(new_delegatee);
+
+		// try transferring the staked amount. This should never fail but if it does, it indicates
+		// bad state and we abort.
+		T::Currency::transfer(
+			new_delegatee,
+			proxy_delegator,
+			stake.total,
+			Preservation::Expendable,
+		)
+		.map_err(|_| Error::<T>::BadState)?;
+
+		// delegate from new delegator to staker.
+		// todo(ank4n) : inline this fn and propagate payee to core staking..
+		Self::accept_delegations(new_delegatee, payee)?;
+
+		Self::delegate(proxy_delegator, new_delegatee, stake.total)?;
+		Self::bond_all(new_delegatee)
+	}
+
+	fn block_delegations(delegatee: &Self::AccountId) -> DispatchResult {
```

Comment:
When a pool is blocked, the root role and bouncer are able to unbond and withdraw members. If I understand correctly, the pool will be calling its `DelegationInterface` when it is blocked / unblocked, keeping them in sync (if we are not removing the pool's blocked state altogether).

Then the nomination pool side will handle `unbond` and `withdraw_unbonded` in the same manner it does now, but on `withdraw_unbonded`, it will destroy the associated ghost account balance and unlock the delegator's corresponding balance back in their account - is this the right assumption?